attorch/layers.py

```diff
@@ -86,296 +87,236 @@ def forward(self, input):
-    def pool(self, p):
-        self.avg.kernel_size = p
-        self.avg.stride = p
+    @property
+    def weight(self):
+        n = self.outdims
+        c, w, h = self.in_shape
+        weight = self.normalized_spatial.expand(n, c, w, h) * self.features.expand(n, c, w, h)
+        weight = weight.view(self.outdims, -1)
+        return weight
 
-    def initialize(self, init_noise=1e-3):
-        # randomly pick centers within the spatial map
-        self.grid.data.uniform_(-.05, .05)
-        self.features.data.fill_(1 / self.in_shape[0])
-        xavier_normal(self.filter.weight.data)
+    def l1(self, average=True):
+        n = self.outdims
+        c, w, h = self.in_shape
+        ret = (self.normalized_spatial.view(self.outdims, -1).abs().sum(1, keepdim=True)
+               * self.features.view(self.outdims, -1).abs().sum(1)).sum()
+        if average:
+            ret = ret / (n * c * w * h)
+        return ret
 
+    def initialize(self, init_noise=1e-3):
+        self.spatial.data.normal_(0, init_noise)
+        self.features.data.normal_(0, init_noise)
         if self.bias is not None:
             self.bias.data.fill_(0)
 
-    def filter_l1(self):
-        return self.filter.weight.abs().mean()
-
-    def feature_l1(self, average=True):
-        if average:
-            return self.features.abs().mean()
-        else:
-            return self.features.abs().sum
-
     def forward(self, x, shift=None):
-        N, c, t, w, h = x.size()
-        feat = self.features.view(1, c, self.outdims)
-
-        if shift is None:
-            grid = self.grid.expand(N * t, self.outdims, 1, 2)
-        else:
-            grid = self.grid.expand(N, self.outdims, 1, 2)
-            grid = torch.stack([torch.clamp(grid + shift[:, i, :][:, None, None, :], -1, 1) for i in range(t)], 1)
-            grid = grid.contiguous().view(-1, self.outdims, 1, 2)
-        z = self.filter(x.contiguous().transpose(2, 1).contiguous().view(-1, c, w, h))
-        z = self.avg(z)
-        y = F.grid_sample(z, grid)
-        y = (y.squeeze(-1) * feat).sum(1).view(N, t, self.outdims)
-
+        N = x.size(0)
+        y = x.view(N, -1) @ self.weight.t()
         if self.bias is not None:
-            y = y + self.bias
+            y = y + self.bias.expand_as(y)
         return y
 
     def __repr__(self):
-        c, _, w, h = self.in_shape
-        r = self.__class__.__name__ + \
-            ' (' + '{} x {} x {}'.format(c, w, h) + ' -> ' + str(self.outdims) + ')'
-        if self.bias is not None:
-            r += ' with bias\n'
-        for ch in self.children():
-            r += '  -> ' + ch.__repr__() + '\n'
-        return r
+        return ('spatial positive ' if self.positive else '') + \
+               ('normalized ' if self.normalize else '') + \
+               self.__class__.__name__ + \
+               ' (' + '{} x {} x {}'.format(*self.in_shape) + ' -> ' + str(
+            self.outdims) + ')'
 
 
-class SpatialTransformerLaplace2d(nn.Module):
```

Comment:
Why did this get removed? This will cause some code to break.

Reply:
Do you use that layer? I don't, but I can put it back.